### PR TITLE
feat: allow `from csv` to accept 4 byte unicode separator chars

### LINKED ISSUE #10132: `from csv` does not accept escaped unicode separators
### Related problem

I work with a lot of CSV files which are separated by the unicode separator '\x1f' https://www.compart.com/en/unicode/U+001F

`from csv` only accepts single characters.

### Describe the solution you'd like

I would like to give `from csv` an escaped sequence of the unicode char.

### Describe alternatives you've considered

I tried to copy it via my shell (wezterm) but this did not work. I am not sure if this the right way (i.e. it this is a terminal emulator issue or a shell issue)

### Additional context and details

_No response_

### PR DESCRIPTION
**Pull request by matthias-Q**
- this PR should close #10132

# Description
* added a flag to `from csv --ascii` that replaces the given `separator with the unicode separator x1f https://www.codetable.net/hex/1f (aka Information Separator One)

# User-Facing Changes
New flags are available for `from csv` ( `--ascii` or short `-a`)

# Tests + Formatting
There are no tests at the moment. Code has been formatted.
- `cargo test --workspace` (breaks with a non related test on my machine)

### CODE REVIEW

**Comment by matthias-Q** [comment]
I have an alternative solution that would be more flexible. The user could enter the unicode sequence (001F) and it would be converted internally into the correct char/u8. I have it running, but I need to do some error handling.

**Review by amtoine — CHANGES_REQUESTED** [review]
i'd wait for the input of some other core team members but i'm not opposed to this :+1: 

maybe having the more flexible version could be great to avoid being stuck with this single escaped sequence :thinking: 

:warning: do you think you could add some test to make sure the issue is fixed by using this new option?

**Comment by matthias-Q** [comment]
@amtoine thanks for you comment. I was also thinking that this definitely needs tests. I will write some once I am done.

Currently, I am a bit stuck with the conversion from the unicode sequence to a correct char. It works fine for some (like x1f or 001F (that's the same)) but not for others.

This is what I currently got:
```rust
    let separator = call
        .get_flag(engine_state, stack, "separator")?
        .map(|v: Value| v.as_string())
        .transpose()?
        .unwrap_or(",".to_string());

    let separator = if separator.len() == 1 {
        separator.chars().next().unwrap_or(',')
    } else if separator.len() == 4 {
        let unicode_sep = u32::from_str_radix(&separator, 16);
        char::from_u32(unicode_sep.unwrap_or(b'\x1f' as u32)).unwrap_or(',')
    } else {
        ','
    };
```

Feels a bit too hacky and defaults to ','. But sometimes we still may want it to fail for some error messages.

Any suggestions? Obviously, this will also not work with escape sequences that are 5 elements long (like emojis)

**Comment by amtoine** [comment]
thanks for the tests :pray: 
(even though the one CSV with `\u001f` as a delimiter looks terrifying :laughing:)

File: crates/nu-command/src/formats/from/csv.rs [hunk 1]
```diff
@@ -136,9 +136,21 @@ fn from_csv(
...
-        .map(|v: Value| v.as_char())
+        .map(|v: Value| -> Result<char, ShellError> {
+            let sep = v.as_string()?;
+            if sep.len() == 1 {
+                Ok(sep.chars().next().unwrap_or(','))
+            } else if sep.len() == 4 {
+                let unicode_sep = u32::from_str_radix(&sep, 16);
+                Ok(char::from_u32(unicode_sep.unwrap_or(b'\x1f' as u32)).unwrap_or(','))
+            } else {
+                // Err(ShellError::NonUtf8Custom("".to_string(), span))
```
[the remark this answers is not in the document]
Comment (amtoine) [reply crates/nu-command/src/formats/from/csv.rs:147]: i think the issue you have here is that you are stuck in a closure and thus cannot `return` the `ShellError` :thinking: 

what about something like the following?
```rust
    let separator = match call.get_flag::<String>(engine_state, stack, "separator")? {
        Some(sep) => {
            if sep.len() == 1 {
                sep.chars().next().unwrap_or(',')
            } else if sep.len() == 4 {
                let unicode_sep = u32::from_str_radix(&sep, 16);
                char::from_u32(unicode_sep.unwrap_or(b'\x1f' as u32)).unwrap_or(',')
            } else {
                return Err(ShellError::NonUtf8Custom(
                    "separator should be a single char or a 4-byte unicode".to_string(),
                    call.span(),
                ));
            }
        }
        None => ',',
    };
```

i did not test it, just made sure it's valid formatted Rust :relieved:

[the remark this answers is not in the document]
Comment (amtoine) [reply crates/nu-command/src/formats/from/csv.rs:147]: to me, it even looks a bit simpler than the calls to `Option::map` and `Result::transpose` :yum:

[the remark this answers is not in the document]
Comment (matthias-Q) [reply crates/nu-command/src/formats/from/csv.rs:147]: Hmm, this won't work since the if/else branches return different types (char and err)

**Comment by matthias-Q** [comment]
Ah yeah, these separators are useful for technical exchanges of CSVs. We use them quite frequently actually. 

I will try to do some follow up PRs for writing and Dataframes. I also find it a bit odd that the flags are called differently among `dfr` and `from` (separator/delimiter)

**Comment by fdncred** [comment]
if i read through this right, we're not using csv's ascii function any longer. maybe we should update the title of the pr sometime. 

i'm happy with this pr and seems to have been simplified. nice work!

**Comment by matthias-Q** [comment]
@fdncred now that I think about it, all of the other chars (`comment`, `quote`, `escape`) could also be unicode. I might wrap this conversion code into a function and reuse it.

I will also add the same functionality to dataframes and to the write code in the next steps, if that's ok.

**Comment by fdncred** [comment]
> I will also add the same functionality to dataframes and to the write code in the next steps, if that's ok.

Sounds good. Do you want to do that in a separate PR or this PR?

**Comment by matthias-Q** [comment]
I would do it in an extra PR

**Comment by fdncred** [comment]
@amtoine Looks like your changes are integrated. Are you ready to land this?

File: crates/nu-command/src/formats/from/csv.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (amtoine) [reply crates/nu-command/src/formats/from/csv.rs]: thanks for handling the error in the end :relieved:

**Comment by amtoine** [comment]
> @amtoine Looks like your changes are integrated. Are you ready to land this?
>
yes the missing error handling has been addressed in the thread above :pray: 

thanks @matthias-Q for the efforts and the contribution :blush: 

let's go ahead and see how this goes :partying_face:

**Pull request merged**